File: node/src/sensor/lock.ts
import { Sensor, SensorCategory, SensorType } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/** Lock states. */
export enum LockState {
  /** Locked. */
  Secured = 0,
  /** Unlocked. */
  Unsecured = 1,
  /** State cannot be determined, e.g. while a motorized lock is moving. */
  Unknown = 2,
}

/**
 * Property names of a lock control.
 *
 * @internal
 */
export enum LockProperty {
  /** The actual current state of the lock. */
  CurrentState = 'currentState',
  /** The desired target state (set by user, transitions to currentState). */
  TargetState = 'targetState',
}

/**
 * Property values of a lock control.
 *
 * @internal
 */
export interface LockControlProperties {
  [LockProperty.CurrentState]: LockState;
  [LockProperty.TargetState]: LockState;
}

/** Read-only proxy interface for a lock control. */
export interface LockControlLike extends SensorLike {
  readonly type: SensorType.Lock;
  readonly onPropertyChanged: Observable<PropertyChangeOf<LockControlProperties>>;

  getValue(property: LockProperty.CurrentState): LockState | undefined;
  getValue(property: LockProperty.TargetState): LockState | undefined;
  getValue(property: string): unknown;
}

/**
 * Lock control. Override `setTargetState()` to drive hardware and call
 * `await super.setTargetState(value)` once the hardware confirms. The base
 * implementation updates both `targetState` and `currentState` to the new value.
 *
 * For asymmetric flows (long-running unlock with intermediate state) override
 * `setTargetState` and write `currentState` separately when transitions complete.
 */
export class LockControl<TStorage extends object = Record<string, any>> extends Sensor<LockControlProperties, TStorage, string> {
  readonly type = SensorType.Lock;
  readonly category = SensorCategory.Control;

  constructor(name = 'Lock', options?: SensorOptions) {
    super(name, options);

    this._writeState({
      [LockProperty.CurrentState]: LockState.Secured,
      [LockProperty.TargetState]: LockState.Secured,
    });
  }

  get currentState(): LockState {
    return this.props.currentState;
  }

  get targetState(): LockState {
    return this.props.targetState;
  }

  /**
   * Set the target state. Override to drive hardware and call
   * `await super.setTargetState(value)` after success. The base implementation
   * syncs both `targetState` and `currentState` to the new value.
   *
   * @param value - Desired lock state from the {@link LockState} enum.
   *
   * @example
   * ```ts
   * import { LockState } from '@camera.ui/sdk';
   * await lock.setTargetState(LockState.Secured);
   * ```
   */
  async setTargetState(value: LockState): Promise<void> {
    this._writeState({
      [LockProperty.TargetState]: value,
      [LockProperty.CurrentState]: value,
    });
  }

  /**
   * Publish the actual lock state. Use it when the physical state diverges from the
   * requested target: motorized locks that take time to rotate (publish `Unknown`
   * while moving), or hardware reporting an out-of-band state change. Read-only from
   * cross-process consumers (`updateValue` ignores it).
   *
   * @param value - Current physical lock state from the {@link LockState} enum.
   *
   * @example
   * ```ts
   * import { LockState } from '@camera.ui/sdk';
   * lock.setCurrentState(LockState.Unknown);
   * ```
   */
  setCurrentState(value: LockState): void {
    this._writeState({ [LockProperty.CurrentState]: value });
  }

  /**
   * Routes generic property writes to the semantic setters. Only `targetState` is externally writable, `currentState` is observed-only.
   *
   * @internal
   */
  async updateValue(property: string, value: unknown): Promise<void> {
    if ((property as LockProperty) === LockProperty.TargetState) {
      await this.setTargetState(value as LockState);
    }
  }
}

/** Registry metadata for {@link LockControl}. */
export const lockMeta = defineSensor({
  type: SensorType.Lock,
  category: SensorCategory.Control,
  assignmentKey: 'lock',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [LockProperty.CurrentState]: {
      type: 'enum',
      values: { secured: LockState.Secured, unsecured: LockState.Unsecured, unknown: LockState.Unknown },
    },
    [LockProperty.TargetState]: {
      type: 'enum',
      values: { secured: LockState.Secured, unsecured: LockState.Unsecured },
      writable: true,
    },
  },
  shortcutable: true,
  cascadeTrigger: { property: LockProperty.CurrentState, value: 1, sustained: true },
  virtual: { properties: { [LockProperty.CurrentState]: 0, [LockProperty.TargetState]: 0 } },
  semantics: {
    domain: SensorDomain.Lock,
    stateProperty: LockProperty.CurrentState,
    commandProperty: LockProperty.TargetState,
    states: { locked: LockState.Secured, unlocked: LockState.Unsecured },
  },
});

File: python/camera_ui_sdk/sensor/lock.py
from __future__ import annotations

from collections.abc import Mapping
from enum import IntEnum, StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class LockState(IntEnum):
    """Lock states."""

    Secured = 0
    """Locked."""
    Unsecured = 1
    """Unlocked."""
    Unknown = 2
    """State cannot be determined, e.g. while a motorized lock is moving."""


class LockProperty(StrEnum):
    """Property names of a lock control."""

    CurrentState = "currentState"
    """The actual current state of the lock."""
    TargetState = "targetState"
    """The desired target state (set by user, transitions to currentState)."""


class LockControlProperties(TypedDict):
    """Property values of a lock control."""

    currentState: int
    targetState: int


class LockPropertyChangeData(TypedDict):
    """Property change payload emitted on LockControlLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a LockProperty value."""
    value: LockState
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class LockControlLike(SensorLike, Protocol):
    """Read-only proxy interface for a lock control."""

    @property
    def type(self) -> SensorType:
        return SensorType.Lock

    @property
    def onPropertyChanged(self) -> Observable[LockPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[LockProperty.CurrentState]) -> LockState | None: ...
    @overload
    def getValue(self, property: Literal[LockProperty.TargetState]) -> LockState | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class LockControl(Sensor[LockControlProperties, TStorage, str], Generic[TStorage]):
    """Lock control.

    Override `setTargetState()` to drive hardware and call
    `await super().setTargetState(value)` once the hardware confirms. The base
    implementation updates both `targetState` and `currentState` to the new value.

    For asymmetric flows (long-running unlock with intermediate state) override
    `setTargetState` and write `currentState` separately when transitions complete.
    """

    _requires_frames = False

    def __init__(self, name: str = "Lock", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state(
            {
                LockProperty.CurrentState.value: int(LockState.Secured),
                LockProperty.TargetState.value: int(LockState.Secured),
            }
        )

    @property
    def type(self) -> SensorType:
        return SensorType.Lock

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Control

    @property
    def currentState(self) -> LockState:
        value = self.props.currentState
        return LockState(value) if value is not None else LockState.Secured

    @property
    def targetState(self) -> LockState:
        value = self.props.targetState
        return LockState(value) if value is not None else LockState.Secured

    async def setTargetState(self, value: LockState) -> None:
        """Set the target state. Override to drive hardware and call
        `await super().setTargetState(value)` after success. The base implementation
        syncs both `targetState` and `currentState` to the new value.

        Args:
            value: Desired lock state from the ``LockState`` enum.

        Example:
            ```python
            from camera_ui_sdk import LockState

            await lock.setTargetState(LockState.Secured)
            ```
        """
        self._write_state(
            {
                LockProperty.TargetState.value: int(value),
                LockProperty.CurrentState.value: int(value),
            }
        )

    def setCurrentState(self, value: LockState) -> None:
        """Publish the actual lock state. Use it when the physical state diverges
        from the requested target: motorized locks that take time to rotate
        (publish ``Unknown`` while moving), or hardware reporting an out-of-band
        state change. Read-only from cross-process consumers (``updateValue``
        ignores it).

        Args:
            value: Current physical lock state from the ``LockState`` enum.

        Example:
            ```python
            from camera_ui_sdk import LockState

            lock.setCurrentState(LockState.Unknown)
            ```
        """
        self._write_state({LockProperty.CurrentState.value: int(value)})

    async def updateValue(self, property: str, value: Any) -> None:
        """Routes generic property writes to the semantic setters.

        Only targetState is externally writable, currentState is observed-only.
        """
        if property == LockProperty.TargetState.value:
            await self.setTargetState(LockState(value))
